docs/fabric_mod_json_spec.md

```diff
@@ -0,0 +1,130 @@
+- **sources**: Project source code repository. Must be a valid URL - it can, however, be a specialized URL for a given VCS (such as Git or Mercurial).
+
+There are no mandatory keys.
+
+The list is not exhaustive - mods may provide additional, non-standard keys (such as **discord**, **slack**, **twitter**) - if possible, they should be valid URLs.
+
+#### EntrypointContainer
+
+An EntrypointContainer is an object.
+
+The keys match the getEntrypoints() "type" field, and are the type of the entrypoints to be listed - "main", "client", "server". The values of those keys are arrays, containing either strings (of the object key "value"'s value) or objects of the following form:
+
+- "adapter": Optional key denoting the language adapter to use. If empty, assume "default".
+- "value": The default language adapter uses this specific key as a string value of any of the following formats:
+  - "my.package.MyClass", which points to a class to be instantiated,
+  - "my.package.MyClass::thing", which points to a static field (contents returned) or method handle (for interface types, proxied automatically) named "thing".
+
+Fabric, by default, will run all "main" entrypoints (type ModInitializer), and all "client" (type ClientModInitializer) or "server" (type DedicatedServerModInitializer) entrypoints after that.
+
+#### NestedJarEntry
+
+An object with the following keys, of which only "file" is mandatory:
+
+- "file" - a string value pointing to a path from the root of the mod to a nested JAR which should be loaded alongside the outer mod JAR.
+
+#### Person
+
+Can be in one of two forms:
+
+- a string (assumed to resolve to a "name" key in the object),
+- an object.
+
+In the case of an object, the following keys are defined:
+
+- **name**: The real name, or username, of the person. Mandatory.
+- **contact** An optional ContactInformation object containing contact information pertaining to the person.
+
+#### VersionRange
+
+A string or array of string declaring supported version ranges. In the case of an array, an "OR" relationship is assumed - that is, only one range has to match for the collective range to be satisfied.
+
+In the case of all versions, `*` is a special string declaring that any version is matched by the range. In addition, exact string matches must be possible regardless of the version type.
+
+For semantic versions, the specification follows a rough subset of the [NPM semver](https://docs.npmjs.com/misc/semver) specification, in particular the following features are supported:
+
+- `=` as a leading character, or the lack of one, denoting an exact match,
+- Version ranges - a set of space-delimited comparators of the `>=`, `>`, `<=`, `<`, `=` or no prefix, following NPM behaviour and declaring an intersection of supported versions within the scope of the string,
+- X-Ranges,
+- Tilde Ranges,
+- Caret Ranges.
+
+### Mandatory fields
+
+- **id**: Contains the mod identifier - a string value matching the `^[a-z][a-z0-9-_]{1,63}$` pattern.
```

Comment:
Doesn't loader require mod ids are at least 3 characters?

Reply:
No. In fact, the error messages from the parsers for missing id strings are incorrect: Loader only prevents `== 1` and `> 64`.

https://github.com/FabricMC/fabric-loader/blob/134386bd5c2fc56e1c397bfd216077a67721cf99/src/main/java/net/fabricmc/loader/discovery/ModResolver.java#L498-L502

Reply:
Huh I see.